src/Compilers/CSharp/Portable/Symbols/DelegateCaches/ModuleScopedDelegateCacheContainer.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.Cci;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    using FieldSymbolsCollection = System.Collections.Concurrent.ConcurrentDictionary<MethodSymbol, ModuleScopedDelegateCacheContainerField>;
+
+    /// <summary>
+    /// This symbol is created while lowering, and is collected and have name assigned before emit.
+    /// </summary>
+    internal class ModuleScopedDelegateCacheContainer : DelegateCacheContainer, IComparer<ModuleScopedDelegateCacheContainerField>
+    {
+        // The container is a top level type, and the compilation can be parallel.
+        // To ensure simple & deterministic output, we need a unique and deterministic sort key when it's created.
+        // When all methods are compiled thus all containers are created, and when things are happening serially before emit, 
+        // we sort the containers by their sort key, then use the indices to name them.
+        private string _name;
+        public override string Name => _name;
```

Comment:
Even at the expense of the debugging experience. Would assert cause problems? Perhaps there is a attribute we can apply to prevent automatic evaluation of this property.